From cf09e8fca1efecc513b3cc9489f59601b58c4fe2 Mon Sep 17 00:00:00 2001
From: bengris32 <bengris32@protonmail.ch>
Date: Sat, 9 Mar 2024 23:15:05 +0000
Subject: [PATCH 1/2] Aperture: Enable MediaTek HFPS Mode for 60 FPS video
 recording

MediaTek devices need to pass this parameter in order to actually record 60 FPS videos.

Tested on Xiaomi 12T, POCO X4 GT and OnePlus Nord 3 5G.

Change-Id: I1e027dcacc48f139c067e4e3ae1d5750d29e65a5
---
 .../aperture/ext/CaptureRequestOptionsBuilder.kt    | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/app/src/main/java/org/lineageos/aperture/ext/CaptureRequestOptionsBuilder.kt b/app/src/main/java/org/lineageos/aperture/ext/CaptureRequestOptionsBuilder.kt
index 4cff361..d4613b1 100644
--- a/app/src/main/java/org/lineageos/aperture/ext/CaptureRequestOptionsBuilder.kt
+++ b/app/src/main/java/org/lineageos/aperture/ext/CaptureRequestOptionsBuilder.kt
@@ -19,6 +19,10 @@ import org.lineageos.aperture.models.NoiseReductionMode
 import org.lineageos.aperture.models.ShadingMode
 import org.lineageos.aperture.models.VideoStabilizationMode
 
+private val FPS60_MTK_KEY_SESSION_PARAMETER = CaptureRequest.Key<Int>(
+    "com.mediatek.streamingfeature.hfpsMode", Int::class.java
+)
+
 @androidx.camera.camera2.interop.ExperimentalCamera2Interop
 fun <ValueT> CaptureRequestOptions.Builder.setOrClearCaptureRequestOption(
     key: CaptureRequest.Key<ValueT>,
@@ -30,11 +34,18 @@ fun <ValueT> CaptureRequestOptions.Builder.setOrClearCaptureRequestOption(
 @androidx.camera.camera2.interop.ExperimentalCamera2Interop
 fun CaptureRequestOptions.Builder.setFrameRate(
     frameRate: FrameRate?
-) = setOrClearCaptureRequestOption(
+) = apply {
+    setOrClearCaptureRequestOption(
     CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
     frameRate?.range?.toRange(),
 )
 
+    // Enable MediaTek HFPS mode if needed
+    if (frameRate == FrameRate.FPS_60) {
+        setOrClearCaptureRequestOption(FPS60_MTK_KEY_SESSION_PARAMETER, 1)
+    }
+}
+
 @androidx.camera.camera2.interop.ExperimentalCamera2Interop
 fun CaptureRequestOptions.Builder.setVideoStabilizationMode(
     videoStabilizationMode: VideoStabilizationMode?
-- 
2.55.0

